### DISCUSSION TITLE
How does tempo difference works?

### DISCUSSION THREAD

**Discussion by korpe1** [wwmm/easyeffects#2924] [General]
I have difficulties to understand, how does the tempo difference in the Pitch effect works?

I mean PipeWire clients (e.g. Firefox) sends audio streams to the PipeWire server in certain tempo/speed and other clients (e.g. EasyEffects) doesn't have ability to affect that (Firefox does what it want). Now when EasyEffect ask that stream from the server: 
- If it start play it at higher tempo the stream will immediately end. You can't play any 1x _live stream_ at 2x speed without this affect.
- If it start play it at lower tempo you have to start keep a buffer, but what when? The buffer will anyway grow too large at the end.

So I think the only way change tempo/speed is that the original source does that.

**Comment by wwmm** [comment]
If I understood your question correctly you are interested in details about how the plugin implementation does what it does. As I did not write the library and I am not familiar with pitch shifting algorithms implementation I can not answer.

In EasyEffects we use the `SoundTouch` library https://surina.net/soundtouch/. They are probably in a better position to explain how it is done.

**Reply by korpe1** [reply]
Thanks for the answer, but I don't wonder how SoundTouch works. It's pretty clear to me and I have used that library alone without EasyEffects. It works for files directly (`soundstretch -tempo=-10% example.mp3`.) and not for audio streams. It's opposite to EasyEffects that is for streams, there's the difference and my questions comes up. I can understand how every player can play audio/video in every speed/tempo they like, but not how EasyEffects, which is just `in -> out` box between two points for the stream, can increase or decrease the speed/tempo. If it increase speed the stream should end immediately (since the source provide the stream in lower speed) and if it decrease speed the stream should start making large buffer (since the source provide the stream in higher speed).

**Reply by wwmm** [reply]
>It's opposite to EasyEffects that is for streams, there's the difference and my questions comes up

It may not seem at first but there is no difference when compared to when you apply soundtouch to a file. Both cases will work with raw pcm audio frames. You can't really process compressed audio. When working with a file the soundtouch executable will have to decompress it and access the pcm audio frames that will look the same as the one passing through a stream. 

It is the same thing in a player. It will apply effects to the pcm frames in a stream and sending it ahead. The fact you have another progam processing these frames at a later stage does not make any difference.

**Reply by korpe1** [reply]
I can understad that the process is the same in both cases, but if we have a audio clip that VLC play and supply to PipeWire server in speed makes it take 10 minutes and at the same time EasyEffects play that stream from the server in speed makes the clip take only 5 minutes, what happens? EasyEffects can't do that because VLC supply audio so slow that EasyEffects hasn't anything to play. The opposite (VLC supply a clip in speed takes 5 minutes and EasyEffects play that stream in speed takes 10 minutes) EasyEffects can do with a buffer, but I think it's hard too eventually. This because after e.g. 2 hours constantly doing this the buffer will be big and it just keeps growing bigger and bigger.

I think that these problems don't be exist only when the source itself chooses speed. For example, if VLC want to play speeder it just readers the file faster or in case HTTP it can download faster or if it want to play slower it does opposite. But for PipeWire client, who take stream from another client and pushes it forward in real time, there are no such options. An effect machine has to work in real time and PipeWire APIs doesn't provide way to ask the source client to change speed.

**Reply by wwmm** [reply]
>>EasyEffects can't do that because VLC supply audio so slow that EasyEffects hasn't anything to play

EasyEffects does not receive audio buffers directly from the audio application but from PipeWire. The player sends to Pipewire that is the one controlling latency as well as the number of buffers per second that is sent to all the clients. It is hard to visualize how your example will actually work because there is the audio server in the middle. The one controlling the pace is not the audio player but the clock PipeWire attaches to the node that drives the effects pipeline. Usually the sound card sink.

**Reply by wwmm** [reply]
>EasyEffects can do with a buffer, but I think it's hard too eventually

EasyEffects does not do buffering. Maybe SoundTouch does it internally. I did not look in detail.

**Reply by wwmm** [reply]
Out of curiosity I did some tests with the pitch plugin. It isn't a plugin I have a reason to use . Something I've notice is that is adds some amount of latency depending on how we configure it.  So it is definitely accumulating data in an internal buffer.

**Reply by korpe1** [reply]
> > EasyEffects can't do that because VLC supply audio so slow that EasyEffects hasn't anything to play
>
> EasyEffects does not receive audio buffers directly from the audio application but from PipeWire. The player sends to Pipewire that is the one controlling latency as well as the number of buffers per second that is sent to all the clients.

Yeah, the PipeWire server has a buffer, but how can it solve this problem? I mean the buffer of the server is for inevitable latencies and isn't for adjust the playback speed. Every PipeWire client negotiate buffer and latencies with the server. That buffer is usually less than a second so it isn't help with this when the speed difference is 5 minutes or more.

> It is hard to visualize how your example will actually work because there is the audio server in the middle. The one controlling the pace is not the audio player but the clock PipeWire attaches to the node that drives the effects pipeline. Usually the sound card sink.

The player doesn't controll the speed of "stream flow", but it definitely handle the playback speed. Or how VLC can always – regardless of hardware – adjust the playback speed from 0.33x to 2.5x?

**Reply by korpe1** [reply]
>>EasyEffects can do with a buffer, but I think it's hard too eventually
>
>EasyEffects does not do buffering. Maybe SoundTouch does it internally. I did not look in detail.

Yeah, EasyEffects is an real-time audio effect machine. And that's why I wonder, how it handle this tempo adjuster? In that comment I just thought the ways it to work.

**Reply by korpe1** [reply]
> Out of curiosity I did some tests with the pitch plugin. It isn't a plugin I have a reason to use . Something I've notice is that is adds some amount of latency depending on how we configure it. So it is definitely accumulating data in an internal buffer.

Do you know, what happens if I use it to decrease the speed for a long time? RAM will running out?
And how this work for increasing the speed? It firstly wait certain time e.g. 10min, then play that buffer 2x speed in 5min and so on?

**Reply by wwmm** [reply]
>Do you know, what happens if I use it to decrease the speed for a long time? RAM will running out?
And how this work for increasing the speed? It firstly wait certain time e.g. 10min, then play that buffer 2x speed in 5min and so on?

Ram won't run out because it isn't just accumulating without sending data ahead. I did not look at the code but it would be bizarre if someone released a library that consumes RAM without bounds.

Based on on my tests yesterday speed increases do not work well if a large percentage is set to the tempo value. It starts to chop sound because of the insertion of silence. Speed increases is probably something that works better when working with files with finite size.
